doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,331 @@
+2. We expose the compiled CJS wrapper function in `context` so that users get to invoke it themselves and skip the default `exports` step.
+
+1 may provide better encapsulation, because with 2 would mean that changes to the CJS wrapper function parameters becomes breaking.
+
+If the module loaded is ESM (`context.format` is `"module"`) all the direct modification to `exports` are no-ops because ESM namespaces are not mutable. Returning a new `exports` for `require(esm)` only affects the caller of `require(esm)`, unless the hook returns a proxy to connect the changes between the new exports object and the module namespace object.
+
+## `link` (import-only): from source code to compiled module records
+
+This needs to work with experimental vm modules for passing module instances around.
+
+Example mock of import-in-the-middle:
+
+```js
+/**
+ * @typedef {{
+ *  specifier: string,
+ *  source: string | Buffer,
+ * }} ModuleLinkContext
+ */
+/**
+ * @typedef {{
+*   module: vm.Module
+* }} ModuleLinkResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleLinkContext} context
+ * @param {(context: ModuleLinkContext) => {ModuleLinkResult}} nextLink
+ * @returns {ModuleLinkResult}
+ */
+function link(url, context, nextLink) {
+  const linked = nextLink(url, context);
+  if (!shouldOverride(url, context.specifier)) {
+    return linked;
+  }
+  assert.strictEqual(linked.module.status, 'linked');  // Original module is linked at this point
+  let source = `import * as original from 'original';`;
+  source += `import { userCallback, name, basedir } from 'util'`;
+  source += `const exported = {}`;
+  for (const key of linked.module.namespace) {
+    source += `let $${key} = original.${key};`;
+    source += `export { $${key} as ${key} }`;
+    source += `Object.defineProperty(exported, '${key}', { get() { return $${key}; }, set (value) { $${key} = value; }});`;
+  }
+  source += `userCallback(exported, name, basedir);`;
+  // This is not yet implemented but should be trivial to implement.
+  linked.module = new vm.SourceTextModuleSync(source, function linkSync(specifier) {
+    if (specifier === 'original') return linked;
+    // Contains a synthetic module with userCallback, name & basedir computed from url
+    if (specifier === 'util') return util;
+  });
+  return linked;
+}
+```
+
+## Invocation order of `resolve` and `load` in ESM
+
+With synchronous hooks, when overriding loading in a way may take a significant amount of time (for example, involving network access), the module requests are processed sequentially, losing the ability to make processing concurrent with a thread pool or event loop, which was what the design of ESM took care to make possible thanks to the `import` syntax. One trick can be used to start concurrent loading requests in `resolve` before blocking for the results in the `load` hook.
```

Comment:
This is very clever, but it feels like a gamble. It seems like _maybe_ the async operation kicked off by `resolve` is resolved by the time `load` runs—but maybe it’s not? I wouldn’t want to recommend this pattern to users and then now event loop turns within the ESM loader become potential breaking changes; like I’d still like to someday be able to refactor the ESM loader to be entirely sync.

Perhaps we can just cut this whole section and consider async activity as a follow-up? It’s covered in #205.

Reply:
This is about concurrency, not about asynchronicity, those are two different things. From a performance perspective, it's beneficial to start all requests early and let them be handled concurrently off-thread. 

> I wouldn’t want to recommend this pattern to users and then now event loop turns within the ESM loader become potential breaking change

The load hooks would still be blocking on the started requests in-order, however between the two another worker can handle the requests concurrently, making it more performant. This pattern doesn't care about event loop turns as long as the resolve and load steps are called in BFS order. Also it matters little what the native event loop of the calling thread does - what matters is what the other thread does. This is basically the user's own event loop re-implemented in JS.

> It’s covered in https://github.com/nodejs/loaders/pull/205.

I assume you are talking about preImport which only applies to top-level imports, but it doesn't address the concurrent ESM module requests, which is what matters for performance of static & synchronous imports.

Reply:
> it doesn’t address the concurrent ESM module requests

I think it does, by the addition of `resolve` and `load` on the `context` passed into `startGraph`. If you look at the HTTPS example in that doc, the intent is that the various network fetches all happen concurrently. Assuming we can implement the hook roughly as described in that design doc, both Node and users would be able to avoid any blocking on async work, whether via `Atomics.wait` or other methods.

Regardless, it seems like this section is just describing a potential section to add to the docs about a particular pattern that users could use, right? If that’s all it is, perhaps we can leave this out for now and reconsider adding it once #205 is implemented, where we can evaluate the various approaches for concurrent async activity to see which one(s) we want to recommend in the docs.

Reply:
> I think it does, by the addition of resolve and load on the context passed into startGraph. If you look at the HTTPS example in that doc, the intent is that the various network fetches all happen concurrently. 

That only addresses dynamic import and root modules, but doesn't handle the case specifically described in this proposal (handling the static imports in `import 'a'; import 'b'; import 'c';` concurrently), whose performance is what matters more in most cases. The section here specifically describes how to handle static imports concurrently while keeping the hooks blocking and synchronous.

I am not sure why this should be removed - isn't this just a place for proposals? It's not nodejs/node?